apps/call_request.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package apps
+
+import (
+	"encoding/json"
+	"io"
+)
+
+// CallRequest envelops all requests sent to Apps.
+type CallRequest struct {
+	// A copy of the Call struct that originated the request. Path and State are
+	// of significance.
+	Call
+
+	// Values are all values entered by the user.
+	Values map[string]interface{} `json:"values,omitempty"`
+
+	// Context of execution, see the Context type for more information.
+	Context Context `json:"context,omitempty"`
+
+	// In case the request came from the command line, the raw text of the
+	// command, as submitted by the user.
+	RawCommand string `json:"raw_command,omitempty"`
+}
+
+// UnmarshalJSON has to be defined since Call is embedded anonymously, and
+// CallRequest inherits its UnmarshalJSON unless it defines its own.
+func (creq *CallRequest) UnmarshalJSON(data []byte) error {
+	// Unmarshal the Call first
+	call := Call{}
+	err := json.Unmarshal(data, &call)
+	if err != nil {
+		return err
+	}
+
+	// Need a type that is just like CallRequest, but without Call to avoid
+	// recursion.
+	structValue := struct {
+		Values     map[string]interface{} `json:"values,omitempty"`
+		Context    Context                `json:"context,omitempty"`
+		RawCommand string                 `json:"raw_command,omitempty"`
+	}{}
+	err = json.Unmarshal(data, &structValue)
+	if err != nil {
+		return err
+	}
+
+	*creq = CallRequest{
+		Call:       call,
+		Values:     structValue.Values,
+		Context:    structValue.Context,
+		RawCommand: structValue.RawCommand,
+	}
+	return nil
+}
+
+func CallRequestFromJSON(data []byte) (*CallRequest, error) {
+	c := CallRequest{}
+	err := json.Unmarshal(data, &c)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+func CallRequestFromJSONReader(in io.Reader) (*CallRequest, error) {
+	c := CallRequest{}
+	err := json.NewDecoder(in).Decode(&c)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+func (creq *CallRequest) GetValue(name, defaultValue string) string {
```

Comment:
It appears convenient to have a function that looks up in `Values`->`State`->default.

@hanzei @mickmister Do you think we should have `GetString`/`GetBool[Value]` with the ^^ semantics defined on `CallRequest`?

Reply:
Filed https://mattermost.atlassian.net/browse/MM-42214